test/integration/dnsrecord/dnsrecord_test.go

```diff
@@ -0,0 +1,488 @@
+	secretAccessKey = flag.String("secret-access-key", "", "AWS secret access key")
+)
+
+func validateFlags() {
+	if len(*accessKeyID) == 0 {
+		panic("need an AWS access key id")
+	}
+	if len(*secretAccessKey) == 0 {
+		panic("need an AWS secret access key")
+	}
+}
+
+var _ = Describe("DNSRecord tests", func() {
+	var (
+		ctx = context.Background()
+
+		logger    *logrus.Entry
+		awsClient *awsclient.Client
+		testEnv   *envtest.Environment
+		mgrCancel context.CancelFunc
+		c         client.Client
+
+		testName = fmt.Sprintf("aws-dnsrecord-it--%s", randomString())
+		zoneName = testName + ".gardener.cloud"
+		zoneID   string
+
+		namespace = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dnsrecord",
+				Namespace: testName,
+			},
+			Data: map[string][]byte{
+				aws.AccessKeyID:     []byte(*accessKeyID),
+				aws.SecretAccessKey: []byte(*secretAccessKey),
+			},
+		}
+		cluster = &extensionsv1alpha1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+	)
+
+	BeforeSuite(func() {
+		repoRoot := filepath.Join("..", "..", "..")
+
+		// enable manager logs
+		logf.SetLogger(zap.New(zap.UseDevMode(true), zap.WriteTo(GinkgoWriter)))
+
+		log := logrus.New()
+		log.SetOutput(GinkgoWriter)
+		logger = logrus.NewEntry(log)
+
+		By("starting test environment")
+		testEnv = &envtest.Environment{
+			CRDInstallOptions: envtest.CRDInstallOptions{
+				Paths: []string{
+					filepath.Join(repoRoot, "example", "20-crd-dnsrecord.yaml"),
+					filepath.Join(repoRoot, "example", "20-crd-cluster.yaml"),
+				},
+			},
+		}
+
+		cfg, err := testEnv.Start()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(cfg).ToNot(BeNil())
+
+		By("setting up manager")
+		mgr, err := manager.New(cfg, manager.Options{
+			MetricsBindAddress: "0",
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(extensionsv1alpha1.AddToScheme(mgr.GetScheme())).To(Succeed())
+		Expect(awsinstall.AddToScheme(mgr.GetScheme())).To(Succeed())
+
+		Expect(dnsrecordctrl.AddToManager(mgr)).To(Succeed())
+
+		var mgrContext context.Context
+		mgrContext, mgrCancel = context.WithCancel(ctx)
+
+		By("starting manager")
+		go func() {
+			defer GinkgoRecover()
+			err := mgr.Start(mgrContext)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+
+		// test client should be uncached and independent from the tested manager
+		c, err = client.New(cfg, client.Options{
+			Scheme: mgr.GetScheme(),
+			Mapper: mgr.GetRESTMapper(),
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(c).NotTo(BeNil())
+
+		flag.Parse()
+		validateFlags()
+
+		awsClient, err = awsclient.NewClient(*accessKeyID, *secretAccessKey, aws.DefaultDNSRegion)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("setting up shoot environment")
+		setupShootEnvironment(ctx, c, namespace, secret, cluster)
+
+		By("creating AWS DNS hosted zone")
+		zoneID = createDNSHostedZone(ctx, awsClient, zoneName)
```

Comment:
Because otherwise I would have to do either of the following:
* Expose one of our existing hosted zone names / IDs in the repo (not really sensitive perhaps, but still internal information), and I didn't want to do it. Also, the account that is configured to perform the integration test would still not have permissions to use my hardcoded zone.
* Introduce an additional TM configuration / secret (with help from the CI team, they would need to change something on their side) to host the zone name / ID. That would perhaps be possible, but I didn't want to do it, creating a temporary zone for the test is simpler from my perspective.

Reply:
Ok, you're right.
After sleeping over it and reading your reasoning, creating a temporary zone indeed sounds like a good idea 😉